Data/ByteString/Internal.hs

```diff
@@ -648,6 +649,19 @@ concat = \bss0 -> goLen0 bss0 bss0
    concat [x] = x
  #-}
 
+-- | Repeats given ByteString n times. More efficient than default definition.
+times :: Integral a => a -> ByteString -> ByteString
+times n (BS fp len) =
+  unsafeCreate (len * (fromIntegral n)) $ \destptr ->
+    withForeignPtr fp $ \p -> 
+      go p destptr (fromIntegral n)
+  where
+    go :: Ptr Word8 -> Ptr Word8 -> Int -> IO ()
+    go _ _ 0 = return ()
+    go fromptr destptr i = do
+      memcpy fromptr destptr len
+      go fromptr (destptr `plusPtr` len) (i-1)
```

Comment:
We can do even better, calling `memcpy` only `log n` times instead of `n`. Check http://hackage.haskell.org/package/vector-0.12.1.2/docs/src/Data.Vector.Generic.Mutable.Base.html#local-6989586621679027630 for an example.

Reply:
Done